server/keyspace/tso_keyspace_group.go

```diff
@@ -0,0 +1,122 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keyspace
+
+import (
+	"context"
+
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/mcs/utils"
+	"github.com/tikv/pd/pkg/storage/endpoint"
+	"github.com/tikv/pd/pkg/storage/kv"
+	"go.uber.org/zap"
+)
+
+// GroupManager is the manager of keyspace group related data.
+type GroupManager struct {
+	ctx context.Context
+	// store is the storage for keyspace group related information.
+	store endpoint.KeyspaceGroupStorage
+}
+
+// NewKeyspaceGroupManager creates a Manager of keyspace group related data.
+func NewKeyspaceGroupManager(ctx context.Context, store endpoint.KeyspaceGroupStorage) *GroupManager {
+	return &GroupManager{
+		ctx:   ctx,
+		store: store,
+	}
+}
+
+// Bootstrap saves default keyspace group info.
+func (m *GroupManager) Bootstrap() error {
+	defaultKeyspaceGroup := &endpoint.KeyspaceGroup{
+		ID: utils.DefaultKeySpaceGroupID,
+		// TODO: define a user kind type
+		UserKind: "default",
+	}
+	err := m.saveKeyspaceGroup(defaultKeyspaceGroup)
+	// It's possible that default keyspace group already exists in the storage (e.g. PD restart/recover),
+	// so we ignore the ErrKeyspaceGroupExists.
+	if err != nil && err != ErrKeyspaceGroupExists {
+		return err
+	}
+
+	return nil
+}
+
+// CreateKeyspaceGroups creates keyspace groups.
+func (m *GroupManager) CreateKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup) error {
+	for _, keyspaceGroup := range keyspaceGroups {
+		// TODO: add replica count
+		kg := &endpoint.KeyspaceGroup{
+			ID:       keyspaceGroup.ID,
+			UserKind: keyspaceGroup.UserKind,
+		}
+		err := m.saveKeyspaceGroup(kg)
+		if err != nil {
+			log.Warn("failed to create keyspace group",
+				zap.Uint32("id", kg.ID),
+				zap.String("use-kind", kg.UserKind),
+				zap.Error(err),
+			)
+		}
+	}
+	return nil
+}
+
+// GetKeyspaceGroups returns all keyspace groups.
+func (m *GroupManager) GetKeyspaceGroups(startID uint32, limit int) ([]*endpoint.KeyspaceGroup, error) {
+	return m.store.LoadKeyspaceGroups(startID, limit)
+}
+
+// GetKeyspaceGroupByID returns the keyspace group by id.
+func (m *GroupManager) GetKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup, error) {
+	var (
+		kg  *endpoint.KeyspaceGroup
+		err error
+	)
+
+	m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+		kg, err = m.store.LoadKeyspaceGroup(txn, id)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	return kg, nil
+}
+
+// DeleteKeyspaceGroupByID deletes the keyspace group by id.
+func (m *GroupManager) DeleteKeyspaceGroupByID(id uint32) error {
+	m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+		return m.store.DeleteKeyspaceGroup(txn, id)
+	})
+	return nil
+}
+
+func (m *GroupManager) saveKeyspaceGroup(keyspaceGroup *endpoint.KeyspaceGroup) error {
+	return m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
```

Comment:
Just wondering, compared to the two round-trips in the change, if the following LoadOrSave operation is pushed down to etcd and just need one round-trip?

```
	txnResp, err := cli.Txn(context.Background()).
		If(clientv3.Compare(clientv3.Version(key), "=", 0)).
		Then(clientv3.OpPut(key, defaultValue)).
		Else(clientv3.OpGet(key)).
		Commit()
```

Reply:
Yes, but the current kv interface doesn't have it.

Reply:
Please ignore this comment then, according to our offline discussion.